morph.go

```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
+	for _, migration := range availableMigrations {
+		// skip if we have the same direction for the migration
+		// we are looking for opposite direction
+		if migration.Direction == direction {
+			continue
+		}
+
+		// we don't have the migration in the map
+		// so we can't rollback it
+		_, ok := migrationsMap[migration.Name]
+		if !ok {
+			continue
+		}
+
+		rollbackMigrations = append(rollbackMigrations, migration)
+	}
+
+	if len(migrations) != len(rollbackMigrations) {
+		return nil, errors.New("not all migrations have opposite migrations")
+	}
+
+	return rollbackMigrations, nil
+}
+
+// GeneratePlan returns the plan to apply these migrations and also includes
+// the safe rollback steps for the given migrations.
+func (m *Morph) GeneratePlan(migrations []*models.Migration) (*models.Plan, error) {
+	rollbackMigrations, err := m.GetOppositeMigrations(migrations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get opposite migrations: %w", err)
+	}
+
+	plan := models.NewPlan(migrations, rollbackMigrations)
+
+	return plan, nil
+}
+
+func (m *Morph) ApplyPlan(plan *models.Plan) error {
+	if err := plan.Validate(); err != nil {
+		return fmt.Errorf("invalid plan: %w", err)
+	}
+
+	revertMigrations := make([]*models.Migration, 0, len(plan.RevertMigrations))
+
+	for i := range plan.Migrations {
+		// add to the revert queue
+		for _, migration := range plan.RevertMigrations {
+			if migration.Name == plan.Migrations[i].Name {
+				revertMigrations = append(revertMigrations, migration)
+				break
+			}
+		}
+
+		err := m.apply(plan.Migrations[i], true)
+		if err != nil {
+			// log the fail step
+			m.config.Logger.Printf("migration %s failed, starting rollback", plan.Migrations[i].Name)
+
+			for j := len(revertMigrations) - 1; j >= 0; j-- {
+				// There is a special case when we are reverting a rollback
+				// We shouldn't save the version if we are trying to restore the last applied migration
+				// here is an example, lets say we have following migrations in the applied migrations table:
+				// migration_1, migration_2, migration_3
+				// Once we initiate the rollback, we will have the following:
+				// migration_3, migration_2, migration_1 (to rollback)
+				// Let's say we have a bug in migration_2 and failed.
+				// We don't remove that version from the database, because migration is not successfully rolled back.
+				// So in this case, we need to apply the migration_2 (up) but it will be in the migrations table.
+				// Therefore we are not saving the version in the database because it will fail on the save version step.
+				skipSave := revertMigrations[j].Direction == models.Up && j == len(revertMigrations)-1
+				rErr := m.apply(revertMigrations[j], !skipSave)
+				if rErr != nil {
+					return fmt.Errorf("could not rollback migrations after trying to migrate: %w", rErr)
+				}
+
+				m.config.Logger.Printf("successfully rolled back migration: %s", revertMigrations[j].Name)
+			}
```

Comment:
For better readability, I'd suggest to break the parent loop on error, and note the index till which it ran. Then move this loop outisde.

Reply:
Sure 👍